src/d/actor/d_a_msw.cpp

```diff
@@ -4,52 +4,342 @@
 //
 
 #include "d/actor/d_a_msw.h"
+#include "d/res/res_msw.h"
 #include "d/d_bg_w.h"
+#include "d/d_bg_s_movebg_actor.h"
 #include "d/d_procname.h"
+#include "m_Do/m_Do_ext.h"
+#include "d/d_com_inf_game.h"
+#include "d/d_s_play.h"
+#include "d/actor/d_a_player.h"
 
 /* 00000078-000002D4       .text ride_call_back__FP4dBgWP10fopAc_ac_cP10fopAc_ac_c */
-void ride_call_back(dBgW*, fopAc_ac_c*, fopAc_ac_c*) {
-    /* Nonmatching */
+void ride_call_back(dBgW* bgw, fopAc_ac_c* i_ac, fopAc_ac_c* i_pt) {
+    msw_class* pActor = static_cast<msw_class*>(i_ac);
+
+    cMtx_YrotS(*calc_mtx, -i_ac->current.angle.y);
+
+    cXyz tmp = i_pt->current.pos - i_ac->current.pos;
+    cXyz pos1;
+    MtxPosition(&tmp, &pos1);
+    cXyz pos2;
+    tmp = i_pt->old.pos - i_ac->current.pos;
+    MtxPosition(&tmp, &pos2);
+
+    s16 z = pos1.z * ((g_regHIO.mChild[0].mFloatRegs[0] + 20.0f) / i_ac->scale.z);
+    s16 x = -pos1.x * ((g_regHIO.mChild[0].mFloatRegs[0] + 20.0f) / i_ac->scale.x);
+
+    cLib_addCalcAngleS2(&i_ac->current.angle.x, z, 10, 0x800);
+    cLib_addCalcAngleS2(&i_ac->current.angle.z, x, 10, 0x800);
+
+    f32 dist = (g_regHIO.mChild[0].mFloatRegs[4] + 50.0f) * fabsf(pos1.z - pos2.z);
+    if (pActor->m2BC.x < dist) {
+        pActor->m2BC.x = dist;
+    }
+
+    dist = (g_regHIO.mChild[0].mFloatRegs[4] + 50.0f) * fabsf(pos1.x - pos2.x);
+    if (pActor->m2BC.z < dist) {
+        pActor->m2BC.z = dist;
+    }
+
+    dist = (g_regHIO.mChild[0].mFloatRegs[8] + 5.0f) * fabsf(pos1.x - pos2.x);
+    if (dist > 10.0f && pActor->m2B0.x < dist) {
+        cLib_addCalc2(&pActor->m2B0.x, dist, 1.0f, g_regHIO.mChild[0].mFloatRegs[7] + 1.2f);
+    }
+
+    dist = (g_regHIO.mChild[0].mFloatRegs[8] + 5.0f) * fabsf(pos1.z - pos2.z);
+    if (dist > 10.0f && pActor->m2B0.z < dist) {
+        cLib_addCalc2(&pActor->m2B0.z, dist, 1.0f, g_regHIO.mChild[0].mFloatRegs[7] + 1.2f);
+    }
+
 }
 
 /* 00000310-00000540       .text chain_Draw__FP9msw_class */
-void chain_Draw(msw_class*) {
-    /* Nonmatching */
+void chain_Draw(msw_class* i_this) {
+    s16 angle1;
+    s32 angle2;
+
+    for (int i = 0; i < 4; i++) {
+        cXyz tmp = i_this->m310[i] - i_this->m2E0[i];
+        angle1 = -cM_atan2s(tmp.y, tmp.z);
+
+        // TODO: std::sqrtf(float)
+        angle2 = cM_atan2s(tmp.x, sqrtf(tmp.y * tmp.y + tmp.z * tmp.z));
```

Comment:
Is the sqrtf not just `tmp.absXZ()`?

Reply:
absXZ calls VECSquareMag so it wouldn't match anyway.
I think the team that did the C-style actors never used absXZ since I've seen this `sqrtf(x*x + z*z)` written out manually a lot in them (maybe a macro?)

Reply:
> absXZ calls VECSquareMag so it wouldn't match anyway. I think the team that did the C-style actors never used absXZ since I've seen this `sqrtf(x*x + z*z)` written out manually a lot in them (maybe a macro?)

I find it unlikely that they used a macro since there's not really any benefit to using macros for that.